packages/@glimmer/validator/lib/tracking.ts

```diff
@@ -11,32 +11,19 @@ import { combine, CONSTANT_TAG, isConstTag, validateTag, valueForTag } from './v
  */
 class Tracker {
   private tags = new Set<Tag>();
-  private last: Tag | null = null;
 
   add(tag: Tag) {
-    if (tag === CONSTANT_TAG) return;
+    if (tag === CONSTANT_TAG || this.tags.has(tag)) return;
 
     this.tags.add(tag);
 
     if (import.meta.env.DEV) {
       unwrap(debug.markTagAsConsumed)(tag);
     }
-
-    this.last = tag;
   }
 
   combine(): Tag {
-    let { tags } = this;
-
-    if (tags.size === 0) {
-      return CONSTANT_TAG;
-    } else if (tags.size === 1) {
-      return this.last as Tag;
-    } else {
-      let tagsArr: Tag[] = [];
-      tags.forEach((tag) => tagsArr.push(tag));
-      return combine(tagsArr);
-    }
+    return combine(Array.from(this.tags));
```

Comment:
I think its likely the memory win and most the performance win is actually this change, which we can make independent of the `last` optimization remaining or removing and the `has` check above. Most browsers have fast-path'd array generation from sets when using `Array.from` in my exploration

Reply:
agree, I will create new PR and bench it